src/VirtoCommerce.CartModule.Data/Services/ShoppingCartSearchService.cs

```diff
@@ -41,17 +41,36 @@ public async Task<ShoppingCartSearchResult> SearchCartAsync(ShoppingCartSearchCr
                     repository.DisableChangesTracking();
 
                     var sortInfos = BuildSortExpression(criteria);
-                    var query = BuildQuery(repository, criteria);
-
-                    result.TotalCount = await query.CountAsync();
+                    var query = BuildQuery(repository, criteria);
+
+                    var forceCountQuery = false;
+
                     if (criteria.Take > 0)
                     {
                         var ids = await query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id)
                                          .Select(x => x.Id)
                                          .Skip(criteria.Skip).Take(criteria.Take)
-                                         .ToArrayAsync();
-
+                                         .ToArrayAsync();
+
+                        result.TotalCount = ids.Count();
+                        // This reduces a load of a relational database by skipping cart count query in case of:
+                        // - First page of the carts is reading (Skip is 0)
+                        // - Count of carts in reading result less than Take value.
+                        if (criteria.Skip > 0 || result.TotalCount == criteria.Take)
+
+                        {
+                            forceCountQuery = true;
+                        }
                         result.Results = (await _cartService.GetByIdsAsync(ids, criteria.ResponseGroup)).OrderBy(x => Array.IndexOf(ids, x.Id)).ToList();
+                    }
+                    else

```

Comment:
rename `forceCountQuery` -> `needExecuteCount`
Please remove this extra `else` block by initially set `needExecuteCount = ` criteria.Take > 0